### PR TITLE
Adding Aqua Scanner CLI (scannercli) support.

### PR DESCRIPTION
**Pull request by Mulgish**
Adding support for Aqua Scanner CLI (scannercli). Please note that it is not the same as Aqua Trivy scanner.
Documentation on scannercli is here: https://support.aquasec.com/support/solutions/articles/16000120206

### CODE REVIEW

**Comment by Mulgish** [comment]
@uhafner Do you mind taking a look at this PR please?

**Comment by uhafner** [comment]
Sorry, I don't have spare time this week. Hopefully next week I have some more time...

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    protected void parseJsonObject(final Report report, final JSONObject jsonReport, final IssueBuilder issueBuilder) {
+        JSONArray resources = jsonReport.optJSONArray("resources");
+        if (resources != null) {
+            parseResources(report, resources, issueBuilder);
+        }
+    }
+
+    private void parseResources(final Report report, final JSONArray resources, final IssueBuilder issueBuilder) {
+        for (int i = 0; i < resources.length(); i++) {
+            JSONObject resourceWrapper = (JSONObject) resources.get(i);
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:40]: The other parsers make sure that the object is of correct type. Does it make sense here as well?
```
if (arrayElement instanceof JSONObject)
```

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            if (!resourceWrapper.isNull("vulnerabilities") && !resourceWrapper.isNull("resource")) {
+                JSONObject resource = resourceWrapper.getJSONObject("resource");
+                JSONArray vulnerabilities = resourceWrapper.getJSONArray("vulnerabilities");
+                for (Object vulnerability : vulnerabilities) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:44]: Same here

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                    report.add(convertToIssue(resource, (JSONObject) vulnerability, issueBuilder));
+                }
+            }
+        }
+    }
+
+    private Issue convertToIssue(final JSONObject resource, final JSONObject vulnerability,
+            final IssueBuilder issueBuilder) {
+        final String fileName = resource.isNull("path") ? resource.optString("name", VALUE_NOT_SET)
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:53]: Wouldn't that be the same:
```
resource.optString("path", resource.optString("name", VALUE_NOT_SET))
```

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                .setSeverity(mapSeverity(vulnerability.optString("aqua_severity", "UNKNOWN")))
+                .setMessage(vulnerability.optString("name", "UNKNOWN"))
+                .setDescription(formatDescription(fileName, resource, vulnerability))
+                .buildAndClean();
+    }
+
+    @SuppressFBWarnings("IMPROPER_UNICODE")
+    private Severity mapSeverity(final String string) {
+        if (AQUA_VULNERABILITY_LEVEL_TAG_LOW.equalsIgnoreCase(string)
+                || AQUA_VULNERABILITY_LEVEL_TAG_NEGLIGIBLE.equalsIgnoreCase(string)) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:66]: Would be simpler with `StringContainsUtils.containsAnyIgnoreCase(string, A, B)`

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                : resource.getString("path");
+        return issueBuilder
+                .setFileName(fileName)
+                .setSeverity(mapSeverity(vulnerability.optString("aqua_severity", "UNKNOWN")))
+                .setMessage(vulnerability.optString("name", "UNKNOWN"))
+                .setDescription(formatDescription(fileName, resource, vulnerability))
+                .buildAndClean();
+    }
+
+    @SuppressFBWarnings("IMPROPER_UNICODE")
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:63]: Not required if used in conjunction with `StringContainsUtils.containsAnyIgnoreCase`.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            return Severity.WARNING_HIGH;
+        }
+        else {
+            return Severity.WARNING_HIGH;
+        }
+    }
+
+    private String formatDescription(final String fileName, final JSONObject resource,
+            final JSONObject vulnerability) {
+        return MessageFormat.format(
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:85]: Is there a reason why you don't use `String.format`?

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                "<p><div><b>Resource</b>: {0}</div><div><b>Installed Version:</b> {1}</div><div><b>Aqua Severity:</b> {2}</div>",
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java:86]: I'm fine with coding this in text. But I am wondering why you are using `j2html` in the `Descriptor` but not here, this looks not consistent (while it creates the same result).

File: src/test/java/edu/hm/hafner/analysis/parser/AquaScannerParserTest.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    protected void assertThatIssuesArePresent(final Report report, final SoftAssertions softly) {
+        softly.assertThat(report).hasSize(14);
+
+        softly.assertThat(report.get(0))
+                .hasSeverity(Severity.WARNING_LOW)
+                .hasFileName("unzip")
+                .hasMessage("CVE-2022-0530");
+        softly.assertThat(report.get(8))
+                .hasSeverity(Severity.WARNING_HIGH)
+                .hasFileName("/opt/app/app-runner.jar:org/jose4j/json/JsonHeaderUtil.class")
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/AquaScannerParserTest.java:32]: extract filename to a CONSTANT

File: src/test/java/edu/hm/hafner/analysis/parser/AquaScannerParserTest.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    void shouldHandleEmptyResultsJenkins67296() {
+        Report report = parse("issue67296.json");
+
+        assertThat(report).isEmpty();
+    }
+
+    @Test
+    void brokenInput() {
+        assertThatThrownBy(() -> parse("eclipse.txt")).isInstanceOf(ParsingException.class);
+    }
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/AquaScannerParserTest.java:66]: I really need to create a base class for this code 😉

**Review by uhafner — CHANGES_REQUESTED** [review]
Thank you for your well written PR. Just some small comments, questions and suggestions.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 1 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: The structure is always JSONObject but there is no harm in making an extra check.
Fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 2 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: Fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 3 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: Good spot, fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 4 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: Using containsAnyIgnoreCase. Made code a bit cleaner.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 5 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: Removed.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 6 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: I used same description as in [TrivyParser](https://github.com/jenkinsci/analysis-model/blob/d4e9d1f9cb7130732151cbff69a238c0fa54ba62/src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java#L90), so reused MessageFormat. Changed to use `String.format` now.

File: src/test/java/edu/hm/hafner/analysis/parser/AquaScannerParserTest.java (hunk 8 above)
Comment (Mulgish) [reply src/test/java/edu/hm/hafner/analysis/parser/AquaScannerParserTest.java]: Done.

File: src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java (hunk 7 above)
Comment (Mulgish) [reply src/main/java/edu/hm/hafner/analysis/parser/AquaScannerParser.java]: Changed  to use j2html. 
Original was grabbed from [TrivyParser](https://github.com/jenkinsci/analysis-model/blob/d4e9d1f9cb7130732151cbff69a238c0fa54ba62/src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java#L90) and it was using unclosed `<p>` tag (not sure if it makes any difference).

### Before (String.format):

```html
<p>
<div><b>Resource</b>: bla.txt</div>
<div><b>Installed Version:</b> 1.0.0</div>
<div><b>Aqua Severity:</b> HIGH</div>
<p>detected bad vulnerability</p>
```

image


### After (j2html):

```html
<div><b>Resource: </b>bla.txt</div>
<div><b>Installed Version: </b>1.0.0</div>
<div><b>Aqua Severity: </b>HIGH</div>
<p>detected bad vulnerability</p>
```

image

**Review by uhafner — APPROVED** [review]

**Pull request merged**